Workflows/Rails upgrades.md

```diff
@@ -0,0 +1,286 @@
+- check the [Rails changelogs](#2-rails-x-upgrade) ahead of time and triage which breaking changes are likely to affect the estimate (e.g. the introduction of Zeitwerk autoloading might require you to rename modules and classes)
+- check in with the rest of the Rails team and ask very specific questions
+  - has anyone performed **this exact** upgrade?
+  - what was the initial estimation and actual time spent on the upgrade?
+  - was the work deployed in phases?
+  - who performed the testing and to what extent?
+  - was the project similar in size or requirements (big/small, internal/client)?
+  - did we discover any gotchas?
+- do a trial upgrade of the Gemfile if possible and note the count of deprecation warnings or application errors when running specs
+- check the dependency project health, projects which haven't been updated for a long time will likely require more work
+- sync with the project team to determine the amount of manual testing you want to do in each specific phase (e.g. testing yourself, small/complete smoke tests by the QA team). Communicate separate development and QA testing estimates
+- sync with the devops team to provide a Ruby installation estimate if applicable
+
+Projects in bad health will require extra care in the estimation phase. The following factors may severly (e.g. 3x) impact the estimate
+
+- low test coverage
+- low internal application functionality knowledge
+- being more than 1 major Rails or Ruby version behind
+- incomplete or proprietary deployment pipeline
+- poor codebase health
+- lack of functional specification
+- forked dependencies
+- unmaintained dependencies (e.g. paperclip)
+- the app is not bootable locally
+- there is no staging environment
+
+**Please perform a sanity check with a TL/LE for all major Rails upgrades.**
+
+## Upgrade steps
+The upgrade process can be split into multiple smaller phases. Splitting the work, reviewing and deploying it individually enables us to:
+
+- introduce smaller code changes,
+- observe and review a single change at a time,
+- perform smaller, more targeted smoke tests of the application,
+- more easily plan multiple smaller amounts of work (usually a a day or so per phase) as opposed to a big (weeks) chunk of work,
+- estimate the work with a higher degree of confidence.
+
+The phases noted below should each contain their own task, acceptance criteria and verification steps (e.g. partial or full smoke test). They should be deployed to the production environment **individually** and there should be some grace period between them when required.
+
+All rules have exceptions so please check in with a fellow TL/LE if you won't be following these guidelines due to project specific constraints. Bundling the upgrade steps together does introduce extra risk so we should be explicit and communicate the tradeoffs to stakeholders.
```

Comment:
I would maybe add here that different steps can be bundled in one deploy, but the local development (and commits?) should still follow the above rules so you can clearly see which changes were made

Reply:
Also, maybe write suggestions on what can be bundled together and what definitely shouldn't?